### ISSUE TITLE
Speech to text from Output audio

### ISSUE THREAD

**Issue by rupeshtech** [Azure-Samples/cognitive-services-speech-sdk#1166]
**Desired Situation**

I want to listen to sound coming out of device (laptop) and convert into text real time (without saving to wav file).

**Current Situation**

I am able to capture and save Output audio to wav file using **NAudio**.

**Following is the sample code to save output to wavfile. this code works fine**

    string outputFileName = @"xxxxxx\recievers.wav";
                var capture = new WasapiLoopbackCapture();
                var writer = new WaveFileWriter(outputFileName, capture.WaveFormat);
                capture.DataAvailable += async (s, e) =>
                {
                    if (writer != null)
                    {
                        await writer.WriteAsync(e.Buffer, 0, e.BytesRecorded);
                        await writer.FlushAsync();
                    }
                };
                capture.RecordingStopped += (s, e) =>
                {
                    if (writer != null)
                    {
                        writer.Dispose();
                        writer = null;
                    }
                    capture.Dispose();
                };
                capture.StartRecording();
                Console.WriteLine("Record Started, Press Any key to stop the record");
                Console.ReadLine();
                capture.StopRecording();

**To get real time speech to text** I trying to push buffer to PushAudioStream to following [sample code][1]

    public static async Task RecognitionWithPushAudioStreamAsync()
            {
                // Creates an instance of a speech config with specified subscription key and service region.
                // Replace with your own subscription key and service region (e.g., "westus").
                var config = SpeechConfig.FromSubscription("YourSubscriptionKey", "YourServiceRegion");
    
                var stopRecognition = new TaskCompletionSource<int>();
    
                // Create a push stream
                using (var pushStream = AudioInputStream.CreatePushStream())
                {
                    using (var audioInput = AudioConfig.FromStreamInput(pushStream))
                    {
                        // Creates a speech recognizer using audio stream input.
                        using (var recognizer = new SpeechRecognizer(config, audioInput))
                        {
                            // Subscribes to events.
                            recognizer.Recognizing += (s, e) =>
                            {
                                Console.WriteLine($"RECOGNIZING: Text={e.Result.Text}");
                            };
    
                            recognizer.Recognized += (s, e) =>
                            {
                                if (e.Result.Reason == ResultReason.RecognizedSpeech)
                                {
                                    Console.WriteLine($"RECOGNIZED: Text={e.Result.Text}");
                                }
                                else if (e.Result.Reason == ResultReason.NoMatch)
                                {
                                    Console.WriteLine($"NOMATCH: Speech could not be recognized.");
                                }
                            };
    
                            recognizer.Canceled += (s, e) =>
                            {
                                Console.WriteLine($"CANCELED: Reason={e.Reason}");
    
                                if (e.Reason == CancellationReason.Error)
                                {
                                    Console.WriteLine($"CANCELED: ErrorCode={e.ErrorCode}");
                                    Console.WriteLine($"CANCELED: ErrorDetails={e.ErrorDetails}");
                                    Console.WriteLine($"CANCELED: Did you update the subscription info?");
                                }
    
                                stopRecognition.TrySetResult(0);
                            };
    
                            recognizer.SessionStarted += (s, e) =>
                            {
                                Console.WriteLine("\nSession started event.");
                            };
    
                            recognizer.SessionStopped += (s, e) =>
                            {
                                Console.WriteLine("\nSession stopped event.");
                                Console.WriteLine("\nStop recognition.");
                                stopRecognition.TrySetResult(0);
                            };
    
                            // Starts continuous recognition. Uses StopContinuousRecognitionAsync() to stop recognition.
                            await recognizer.StartContinuousRecognitionAsync().ConfigureAwait(false);
    						var capture = new WasapiLoopbackCapture();
    						var writer = new WaveFileWriter(outputFileName, capture.WaveFormat);
    						capture.DataAvailable += async (s, e) =>
    						{
    							if (writer != null)
    							{
    								await writer.WriteAsync(e.Buffer, 0, e.BytesRecorded);
    								await writer.FlushAsync();
    								pushStream.Write(e.Buffer, e.BytesRecorded); // try to push buffer here
    							}
    						};
    						capture.RecordingStopped += (s, e) =>
    						{
    							if (writer != null)
    							{
    								writer.Dispose();
    								writer = null;
    							}
    							capture.Dispose();
    						};
    						capture.StartRecording();
    						Console.WriteLine("Record Started, Press Any key to stop the record");
    						Console.ReadLine();
    						capture.StopRecording();
                            pushStream.Close();
    
                            // Waits for completion.
                            // Use Task.WaitAny to keep the task rooted.
                            Task.WaitAny(new[] { stopRecognition.Task });
    
                            // Stops recognition.
                            await recognizer.StopContinuousRecognitionAsync().ConfigureAwait(false);
                        }

in short I am opening a pushstream like 

    using (var pushStream = AudioInputStream.CreatePushStream())

and trying to push buffer in the stream like 

    pushStream.Write(e.Buffer, e.BytesRecorded);

**Not getting any speech recognized from Cognitive service.**

Thanks in advance for help

  [1]: https://github.com/Azure-Samples/cognitive-services-speech-sdk/blob/master/samples/csharp/sharedcontent/console/speech_recognition_samples.cs

**Comment by jhakulin** [comment]
@rupeshtech I believe the problem is that NAudio WasapiLoopbackCapture provides audio in 44.1kHz stereo format and using that with AudioInputStream.CreatePushStream() (which expects 16kHz, mono stream by default, see https://docs.microsoft.com/en-us/dotnet/api/microsoft.cognitiveservices.speech.audio.pushaudioinputstream?view=azure-dotnet). 
Could you please check the format NAudio provides and configure pushstream using https://docs.microsoft.com/en-us/dotnet/api/microsoft.cognitiveservices.speech.audio.pushaudioinputstream.-ctor?view=azure-dotnet#Microsoft_CognitiveServices_Speech_Audio_PushAudioInputStream__ctor_Microsoft_CognitiveServices_Speech_Audio_AudioStreamFormat_ constructor.

**Comment by rupeshtech** [comment]
@jhakulin Thanks for your response.
As per you suggestion I am creating pushstream with following code
` using (var pushStream = AudioInputStream.CreatePushStream(AudioStreamFormat.GetWaveFormatPCM(44100,16,1)))`
but still getting **Not getting any speech recognized from Cognitive service.**

**Comment by jhakulin** [comment]
@rupeshtech I looked this more and find out that format is 48kHz, stereo, IEEE FLOAT encoded that comes out from WasapiLoopbackCapture object. Speech SDK expects PCM encoded format, so data needs to be converted from float to pcm.
I modified your code to do that and this example transcribed the loopback audios correctly. Please note the example here may not be the most optimal solution what comes to buffer handing etc. but will give you ideas for next steps.

 ```csharp
        private static byte[] ConvertFloatArrayToInt16ByteArray(float[] data)
        {
            MemoryStream dataStream = new MemoryStream();
            int x = sizeof(Int16);
            Int16 maxValue = Int16.MaxValue;
            int i = 0;
            while (i < data.Length)
            {
                dataStream.Write(BitConverter.GetBytes(Convert.ToInt16(data[i] * maxValue)), 0, x);
                ++i;
            }
            byte[] bytes = dataStream.ToArray();
            dataStream.Dispose();
            return bytes;
        }

        public static async Task RecognitionWithPushAudioStreamAsync()
        {
            // Creates an instance of a speech config with specified subscription key and service region.
            // Replace with your own subscription key and service region (e.g., "westus").
            var config = SpeechConfig.FromSubscription("YourSubscriptionKey", "YourServiceRegion");

            var stopRecognition = new TaskCompletionSource<int>();

            // Create a push stream
            //AudioStreamFormat format = AudioStreamFormat.GetWaveFormatPCM(48000, 16, 2);
            using (var pushStream = AudioInputStream.CreatePushStream(AudioStreamFormat.GetWaveFormatPCM(48000, 16, 2)))
            {
                using (var audioInput = AudioConfig.FromStreamInput(pushStream))
                {
                    // Creates a speech recognizer using audio stream input.
                    using (var recognizer = new SpeechRecognizer(config, audioInput))
                    {
                        // Subscribes to events.
                        recognizer.Recognizing += (s, e) =>
                        {
                            Console.WriteLine($"RECOGNIZING: Text={e.Result.Text}");
                        };

                        recognizer.Recognized += (s, e) =>
                        {
                            if (e.Result.Reason == ResultReason.RecognizedSpeech)
                            {
                                Console.WriteLine($"RECOGNIZED: Text={e.Result.Text}");
                            }
                            else if (e.Result.Reason == ResultReason.NoMatch)
                            {
                                Console.WriteLine($"NOMATCH: Speech could not be recognized.");
                            }
                        };

                        recognizer.Canceled += (s, e) =>
                        {
                            Console.WriteLine($"CANCELED: Reason={e.Reason}");

                            if (e.Reason == CancellationReason.Error)
                            {
                                Console.WriteLine($"CANCELED: ErrorCode={e.ErrorCode}");
                                Console.WriteLine($"CANCELED: ErrorDetails={e.ErrorDetails}");
                                Console.WriteLine($"CANCELED: Did you update the subscription info?");
                            }

                            stopRecognition.TrySetResult(0);
                        };

                        recognizer.SessionStarted += (s, e) =>
                        {
                            Console.WriteLine("\nSession started event.");
                        };

                        recognizer.SessionStopped += (s, e) =>
                        {
                            Console.WriteLine("\nSession stopped event.");
                            Console.WriteLine("\nStop recognition.");
                            stopRecognition.TrySetResult(0);
                        };

                        // Starts continuous recognition. Uses StopContinuousRecognitionAsync() to stop recognition.
                        await recognizer.StartContinuousRecognitionAsync().ConfigureAwait(false);
                        var capture = new WasapiLoopbackCapture();
                        var writer = new WaveFileWriter("c:\\temp\\naudio2.wav", capture.WaveFormat);
                        capture.DataAvailable += async (s, e) =>
                        {
                            if (writer != null)
                            {
                                if (e.BytesRecorded != 0)
                                {
                                    var floatArray = new float[e.BytesRecorded / 4];
                                    Buffer.BlockCopy(e.Buffer, 0, floatArray, 0, e.BytesRecorded);

                                    byte[] ba = ConvertFloatArrayToInt16ByteArray(floatArray);
                                    pushStream.Write(ba); // try to push buffer here
                                }
                            }
                        };
                        capture.RecordingStopped += (s, e) =>
                        {
                            if (writer != null)
                            {
                                writer.Dispose();
                                writer = null;
                            }
                            capture.Dispose();
                        };
                        capture.StartRecording();
                        Console.WriteLine("Record Started, Press Any key to stop the record");
                        Console.ReadLine();
                        capture.StopRecording();
                        pushStream.Close();

                        // Waits for completion.
                        // Use Task.WaitAny to keep the task rooted.
                        Task.WaitAny(new[] { stopRecognition.Task });

                        // Stops recognition.
                        await recognizer.StopContinuousRecognitionAsync().ConfigureAwait(false);
                    }
                }
            }
        }
 ```

**Comment by jhakulin** [comment]
Closing the issue as solution has been provided. Let us know if there is more to help with.
